### PR TITLE
Feat: Add a custom icon

### PR DESCRIPTION
**Pull request by Dedelweiss**
In this PR, I'm adding the new custom icons feature. This allows users to add one or more custom icons (in png or svg format).

- [x] Add custom
- [x] Src from the CSS file
- [x] Fix size for png
- [x] Multiple custom icons
- [x] The user must use a name in lowercase for custom icon 
- [x] Add tests 

### Code:

image

IE:
Dark icon is optional so you can just put:
```"-social-links:custom::https://joyofcoding.org/sponsoring.html::joycoding.png,custom::lucasleblanc.com::lucasLogo.svg,github::https://github.com/Dedelweiss"```

### Result:
Screenshot 2023-06-29 at 16 16 48

I think a next Issue/PR will be needed to improve the tap target of the footer's icons.

Fixes: #16856

### CODE REVIEW

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala [hunk 1]
```diff
@@ -4,11 +4,12 @@ import java.nio.file.Path
...
-  case Github(ghUrl: String) extends SocialLinks(ghUrl, "gh")
-  case Twitter(tUrl: String) extends SocialLinks(tUrl, "twitter")
-  case Gitter(gUrl: String) extends SocialLinks(gUrl, "gitter")
-  case Discord(dUrl: String) extends SocialLinks(dUrl, "discord")
+enum SocialLinks(val url: String, val whiteIcon: String, val darkIcon: String, val className: String):
+  case Github(ghUrl: String) extends SocialLinks(ghUrl, "", "", "gh")
+  case Twitter(tUrl: String) extends SocialLinks(tUrl, "", "", "twitter")
+  case Gitter(gUrl: String) extends SocialLinks(gUrl, "", "", "gitter")
+  case Discord(dUrl: String) extends SocialLinks(dUrl, "", "", "discord")
+  case Custom(cUrl: String, firstIcon: String, secondIcon: String) extends SocialLinks(cUrl, firstIcon, secondIcon, "custom")
```
Comment (KacperFKorban) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:12]: I'm not a fan of adding those fields to the parent class, just so they will be used in one case and empty in all others. I'll suggest something like this:

```suggestion
enum SocialLinks(val url: String, val className: String):
  case Github(ghUrl: String) extends SocialLinks(ghUrl, "gh")
  case Twitter(tUrl: String) extends SocialLinks(tUrl, "twitter")
  case Gitter(gUrl: String) extends SocialLinks(gUrl, "gitter")
  case Discord(dUrl: String) extends SocialLinks(dUrl, "discord")
  case Custom(cUrl: String, lightIcon: String, darkIcon: String) extends SocialLinks(cUrl, "custom")
```

Then you can match on the link in `HtmRenderer.scala`, like so:

```scala
        link match
          case SocialLinks.Custom(_, lightIcon, darkIcon) =>
            Seq(
              img(cls := s"icon-button ${icon(link)}", src := s"../../../../images/$lightIcon"),
              img(cls := s"icon-button ${icon(link)}-dark", src := s"../../../images/$darkIcon")
            )
          case _ =>
            button(cls := s"icon-button ${icon(link)}")
```

Also, it should be `lightIcon`, not `whiteIcon`.

File: scaladoc/src/dotty/tools/scaladoc/renderers/HtmlRenderer.scala [hunk 2]
```diff
@@ -175,7 +175,13 @@ class HtmlRenderer(rootPackage: Member, members: Map[DRI, Member])(using ctx: Do
     def icon(link: SocialLinks) = link.className
     args.socialLinks.map { link =>
       a(href := link.url) (
-        button(cls := s"icon-button ${icon(link)}")
+        if icon(link) == "custom" then
+          Seq(
+            img(cls := s"icon-button ${icon(link)}", src := s"../../../../images/${link.whiteIcon}"),
```
Comment (KacperFKorban) [inline scaladoc/src/dotty/tools/scaladoc/renderers/HtmlRenderer.scala:180]: I think that the resources should be handled in a better way. This relative link was used in the css, since it was resolved as a url from the css file path. Right now it won't be possible to fetch the images if the scaladoc is hosted from a place other than the root folder of the generated scaladoc. 

Though not an expert on that, so maybe @Florian3k might have some insights.

File: scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala [hunk 3]
```diff
@@ -56,7 +56,8 @@ class ScaladocSettings extends SettingGroup with AllScalaSettings:
 
   val socialLinks: Setting[List[String]] =
     MultiStringSetting("-social-links", "social-links",
-      "Links to social sites. '[github|twitter|gitter|discord]::link' syntax is used.")
+      "Links to social sites. '[github|twitter|gitter|discord]::link' syntax is used." +
+        "'custom::link::white_icon_name::black_icon_name' is also allowed, in this case icons must be present in 'images/'' directory.")
```
Comment (KacperFKorban) [inline scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:60]: Maybe rephrase it to sth like:

```suggestion
      "Links to social sites. '[github|twitter|gitter|discord]::link' or 'custom::link::light_icon_file_name[::dark_icon_file_name]' syntax is used. For custom links, the icons must be present in '_assets/images/'"
```

**Comment by Dedelweiss** [comment]
Thank you very much @KacperFKorban for the feedback. Just added the corrections that you sent to me, now I'm looking for a better dynamic src for the images.

File: scaladoc/src/dotty/tools/scaladoc/renderers/HtmlRenderer.scala (hunk 2 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/renderers/HtmlRenderer.scala]: One solution would be to fix the name of the custom logo and to use it in the css like this:

.icon-button.custom::after {
  content: url("../../../../images/icon-buttons/custom/light/default.svg");
}

But when if someone doesn't use a custom, there is an error.

**Comment by ckipp01** [comment]
One more thing to think about here after speaking with you @Dedelweiss about this today. Keep in mind that with the solution that you outlined it will only work for 1 custom icon. So if the user wants two, then it falls apart. I think whatever solution we come up with it should allow for X amount of custom icons/links. Like the number shouldn't matter.

**Comment by Florian3k** [comment]
I'm not a CSS expert, but maybe it's possible to do something like this:
In CSS file add this fragment:
```css
.icon-button.custom::after {
  content: attr(data-custom-icon url);
}
```
And in the renderer add this `data-custom-icon` attribute to the button:
```html
<button data-custom-icon="../../../../path/to/image.svg">
```
This could possibly be extended for more attributes for dark icon etc.

**Comment by Florian3k** [comment]
I just realised that support for this is practically nonexistent: https://caniuse.com/mdn-css_types_attr_type-or-unit_url
I'll try to figure out something else. I may be necessary to move those url's from css to `src` attributes for images in html.

**Comment by Dedelweiss** [comment]
> I just realised that support for this is practically nonexistent: https://caniuse.com/mdn-css_types_attr_type-or-unit_url I'll try to figure out something else. I may be necessary to move those url's from css to `src` attributes for images in html.

Thank you @Florian3k, I will look at it too, see if it's possible to do something different.

File: scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala [hunk 4]
```diff
@@ -43,7 +43,7 @@ class ScaladocSettings extends SettingGroup with AllScalaSettings:
 
   val socialLinks: Setting[List[String]] =
     MultiStringSetting("-social-links", "social-links",
-      "Links to social sites. '[github|twitter|gitter|discord]::link' syntax is used.")
+      "Links to social sites. '[github|twitter|gitter|discord]::link' or 'custom::link::light_icon_file_name[::dark_icon_file_name]' syntax is used. For custom links, the icons must be present in '_assets/images/'")
```
Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]:
```suggestion
      "Links to social sites. '[github|twitter|gitter|discord]::link' or 'custom::link::light_icon_file_name::dark_icon_file_name' syntax is used. For custom links, the icons must be present in '_assets/images/'")
```

This should be this right?

Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: I'll probably get there when I keep reviewing, but what happens when the link isn't there? Does the user get an error message?

Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: Also here we have `custom` and when we match on this we match on `custom`. Should we instead allow them to use the name of their choice? For example let's say I want to use a link to my mastodon account, I may want to do `mastodon::link::light::dark`.

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala [hunk 5]
```diff
@@ -19,5 +20,8 @@ object SocialLinks:
       case "gitter" => Left(errorPrefix + "For 'gitter' arg expected one argument: url")
       case "discord" if splitted.size == 2 => Right(Discord(splitted(1)))
       case "discord" => Left(errorPrefix + "For 'discord' arg expected one argument: url")
+      case "custom" if splitted.size == 4 => Right(Custom(splitted(1), splitted(2), splitted(3)))
+      case "custom" if splitted.size == 3 => Right(Custom(splitted(1), splitted(2), splitted(2)))
```
Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:24]: Ah, so this answers my question from above. If they don't provide all of them, then you just use the icon for both right?

**Review by ckipp01 — CHANGES_REQUESTED** [review]
Just a few nits

File: scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala (hunk 4 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: It's between brackets because the dark icon is optional

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala (hunk 5 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:24]: Yes, that's right, dark mode is optional, so if it's not provided, the icon will be used for both modes.

File: scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala (hunk 4 above)
Comment (ckipp01) [reply scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: O ok, makes sense.

Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: Ho Indeed, this can be useful, and I'm going to try to do it. For example, if I want to put several custom icons, I have this kind of settings :
```social-links:custom::https://www.epfl.com::epfl.png,custom::scala.com::scala_logo.png```
So it can be useful to have:
```social-links:epfl::https://www.epfl.com::epfl.png,scala::scala.com::scala_logo.png```

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala (hunk 5 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:24]: Example: 
```social-links:custom::https://www.epfl.com::epfl.png,custom::scala.com::scala_logo.png``` works like
```social-links:custom::https://www.epfl.com::epfl.png::epfl.png,custom::scala.com::scala_logo.png::scala_logo.png```

File: scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala (hunk 4 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/ScaladocSettings.scala:46]: Hello Chris! In my last commit, I added a way for the user to use a name for their custom icon using a regex. It's a first idea and I'm very interested if you have any corrections, other ideas or suggestions.

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala [hunk 6]
```diff
@@ -5,11 +5,15 @@ enum SocialLinks(val url: String, val className: String):
   case Twitter(tUrl: String) extends SocialLinks(tUrl, "twitter")
   case Gitter(gUrl: String) extends SocialLinks(gUrl, "gitter")
   case Discord(dUrl: String) extends SocialLinks(dUrl, "discord")
+  case Custom(cUrl: String, lightIcon: String, darkIcon: String) extends SocialLinks(cUrl, "custom")
 
 object SocialLinks:
+  val LowercaseNamePattern = "^[a-z]+$".r
```
Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:11]: I think since we're introducing some more parsing here do you think you could add a small unit test to just ensure the logic for `parse` works as expected?

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala [hunk 7]
```diff
@@ -19,5 +23,8 @@ object SocialLinks:
       case "gitter" => Left(errorPrefix + "For 'gitter' arg expected one argument: url")
       case "discord" if splitted.size == 2 => Right(Discord(splitted(1)))
       case "discord" => Left(errorPrefix + "For 'discord' arg expected one argument: url")
+      case LowercaseNamePattern() if splitted.size == 4 => Right(Custom(splitted(1), splitted(2), splitted(3)))
+      case LowercaseNamePattern() if splitted.size == 3 => Right(Custom(splitted(1), splitted(2), splitted(2)))
+      case LowercaseNamePattern() => Left(errorPrefix + "For 'custom' two minimum arguments are expected: url, white icon name, [dark icon name]")
```
Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:28]: Nice, I think this is a lot nicer than what we had before with `custom`.

**Review by ckipp01 — CHANGES_REQUESTED** [review]
Just one nit on adding a test, but other than that this should be good to go.

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala (hunk 7 above)
Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:28]: Yes, thank you! Do you think I should write the pattern as custom() to specify that it's for custom icons or leave it like this to specify that it's a pattern for lowercase names?

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala [hunk 8]
```diff
@@ -5,11 +5,15 @@ enum SocialLinks(val url: String, val className: String):
...
+  case Custom(cUrl: String, lightIcon: String, darkIcon: String) extends SocialLinks(cUrl, "custom")
 
 object SocialLinks:
+  val LowercaseNamePattern = "^[a-z]+$".r
+
   def parse(s: String): Either[String, SocialLinks] =
     val errorPrefix = s"Social links arg $s is invalid: "
     val splitted = s.split("::")
+
     splitted.head match {
```
Comment (ckipp01) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:17]: What do you think about maybe adding `.toLowerCase` here? I think it might be annoying if someone passes in `Custom` and it fails because it wants `custom`.

Comment (Dedelweiss) [reply scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:17]: Hello Chris, ho I hadn't thought of that, but it's much better. I'll test it to see if it's not a problem elsewhere.

**Review by ckipp01 — APPROVED** [review]
This looks good to me! Thanks @Dedelweiss!

File: scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala (hunk 7 above)
Comment (KacperFKorban) [inline scaladoc/src/dotty/tools/scaladoc/SocialLinks.scala:28]: It should be `light` instead of `white`. Also, the sentence seems a bit off.
I suggest: `For the 'custom' link, a minimum of two arguments is expected: URL, light icon file name, [dark icon file name]`

**Review by KacperFKorban — COMMENTED** [review]
Looks good now, one last nitpick from me.

**Review by KacperFKorban — APPROVED** [review]
